crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -53,94 +52,147 @@ pub fn get_matrix_from_inst(py: Python, inst: &PackedInstruction) -> PyResult<Ar
+    fn from_qubit(n: usize, versor: VersorU2) -> Self {
+        let mut qubits: [VersorSU2; 2] = Default::default();
+        qubits[n] = versor.su2;
+        Self {
+            phase: versor.phase,
+            qubits,
+        }
+    }
+
+    /// Apply a new gate to one of the two qubits.
+    #[inline]
+    fn apply_on_qubit(&mut self, n: usize, versor: &VersorU2) {
+        self.phase += versor.phase;
+        self.qubits[n] = versor.su2 * self.qubits[n];
+    }
+
+    /// Construct the two-qubit gate matrix implied by these two runs.
+    #[inline]
+    fn matrix_into<'a>(&self, target: &'a mut [[Complex64; 4]; 4]) -> ArrayView2<'a, Complex64> {
+        let q0 = VersorU2 {
+            phase: self.phase,
+            su2: self.qubits[0],
+        }
+        .matrix_contiguous();
+
+        // This is the manually unrolled Kronecker product.
+        let [iz, iy, ix, scalar] = *self.qubits[1].0.quaternion().coords.as_ref();
+        let q1_row = [Complex64::new(scalar, iz), Complex64::new(iy, ix)];
+        for out_row in 0..2 {
+            for out_col in 0..4 {
+                target[out_row][out_col] = q1_row[(out_col & 2) >> 1] * q0[out_row][out_col & 1];
+            }
+        }
+        let q1_row = [Complex64::new(-iy, ix), Complex64::new(scalar, -iz)];
+        for out_row in 0..2 {
+            for out_col in 0..4 {
+                target[out_row + 2][out_col] =
+                    q1_row[(out_col & 2) >> 1] * q0[out_row][out_col & 1];
+            }
+        }
+        aview2(target)
+    }
+}
+
+/// Extract a versor representation of an arbitrary 1q DAG instruction.
+fn versor_from_1q_gate(py: Python, inst: &PackedInstruction) -> PyResult<VersorU2> {
+    let versor_result = if let Some(gate) = inst.standard_gate() {
+        VersorU2::from_standard(gate, inst.params_view())
+    } else {
+        VersorU2::from_ndarray(&get_matrix_from_inst(py, inst)?.view(), 1e-12)
+    };
+    versor_result.map_err(|err| QiskitError::new_err(err.to_string()))
+}
+
 /// Return the matrix Operator resulting from a block of Instructions.
+///
+/// # Panics
+///
+/// If any node in `op_list` is not a 1q or 2q gate.
 pub fn blocks_to_matrix(
     py: Python,
     dag: &DAGCircuit,
     op_list: &[NodeIndex],
     block_index_map: [Qubit; 2],
 ) -> PyResult<Array2<Complex64>> {
-    let map_bits = |bit: &Qubit| -> u8 {
-        if *bit == block_index_map[0] {
-            0
+    #[derive(Clone, Copy, PartialEq, Eq)]
+    enum Qarg {
+        Q0 = 0,
+        Q1 = 1,
+        Q01,
+        Q10,
+    }
+    let qarg_lookup = |qargs| {
+        let interned = dag.get_qargs(qargs);
+        if interned.len() == 1 {
+            if interned[0] == block_index_map[0] {
+                Qarg::Q0
+            } else {
+                Qarg::Q1
+            }
+        } else if interned.len() == 2 {
+            if interned[0] == block_index_map[0] {
+                Qarg::Q01
+            } else {
+                Qarg::Q10
+            }
```

Comment:
Not that I think it matters very much, but I feel like we could probably maybe map the value of `qargs` directly to `Q0`, `Q1`, `Q01`, and `Q10` and avoid the interner lookup. There should only be 4 interned ids in the context of a 2q block so it feels like once we see an id once we don't need to do the full get_qargs() call on every instruction. But I highly doubt this is a bottlneck at all, so probably not worth the extra complexity.

Reply:
Done in 3bfb8f7428, though I don't think it has much impact on the timings.